rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+instructions:
+
+```python
+# Sample tile sizes
+i_tiles = sch.sample_perfect_tile(i, n=4)
+j_tiles = sch.sample_perfect_tile(j, n=4)
+k_tiles = sch.sample_perfect_tile(k, n=2)
+# Tile the loops according to the random variables
+i_0, i_1, i_2, i_3 = sch.split(loop=i, factors=i_tiles)
+j_0, j_1, j_2, j_3 = sch.split(loop=j, factors=j_tiles)
+k_0, k_1           = sch.split(loop=k, factors=k_tiles)
+# Organize the loops into "SSRSRS" 6-level tiles
+sch.reorder(
+    i_0, j_0, # S: the 1st spatial tile
+    i_1, j_1, # S: the 2nd spatial tile
+    k_0,      # R: the 1st reduction tile
+    i_2, j_2, # S: the 3rd spatial tile
+    k_1,      # R: the 2nd reduction tile
+    i_3, j_3, # S: the 4th spatial tile
+)
+```
+
+### 3.4. AutoScheduler-style Design Space Generation
+
+AutoScheduler (Ansor) generates schedule templates by applying their SearchRules to each stage.
+SearchRule analyzes TE and eagerly trigger schedule primitives accordingly in its internally
+maintained mini IR.
+
+As introduced in Section 3.2, composite schedule rules are equivalent to AutoScheduler's SearchRule
+in TensorIR scheduling. To further generate a design space for scheduling, sampling instructions are
+used in composite schedule rules. Similarly, the sketch generation phase in AutoScheduler is
+equivalent to applying composite schedule rules to each block in TensorIR.
+
+Several built-in composite schedule rules are shipped with our system to align with Ansor's design
+space:
+
+* Multi-level tiling
+* Inline pure spatial blocks
+* Parallelize & vectorize & unroll
+* Auto tensorize
+
+Developers may implement their own rules in either Python or C++. They may specify which rules to
+use with the following syntax:
+
+```python
+from tvm import meta_schedule as ms
+
+design_space_generator = ms.PostOrderApply(rules=[
+    ms.MultiLevelTiling(...),
+    CustomRule(...),
+    ms.OtherBuiltinRules(...),
+])
+
+```
+
+### 3.5. Unifying manual schedule / AutoTVM / Ansor
+
+In this section, we show that the design space induced by TE manual schedule, AutoTVM and Ansor are
+all subsets of meta schedule, and meta schedule further allows mixing those three styles to search
+jointly.
+
+**Manual schedule**. The TE schedule is a special case of a meta schedule program, where there is no
+randomness introduced by sampling instructions. It is a single point in terms of design space.
+
+**AutoTVM (Template-based tuning)**. It is more natural representation of AutoTVM’s schedule
+templates (knobs) by writing one or more schedule functions in meta schedule with sampling
+instructions. The execution traces generated by the schedule functions are the design space to be
+explored.
+
+**AutoScheduler (Ansor, Template-free tuning)**. As mentioned in the previous section, application
+of composite schedule rules generates the design space, which is equivalent to Ansor’s sketch
+generation.
+
+**Mixing styles in design space definition**. By taking union of the spaces induced by the three
```

Comment:
Yeah I agree. The RFC itself is more related to "a DSL that defines a search space", not task extraction / lowering (i suppose it is part of TECompiler?), but it will be helpful to demonstrate the entire search process

Reply:
I will leave step 1 & 2 to TECompiler. Step 3 & 5 are described in our Section 4.2 (exploring the design space)